src/tools/ceph_dedup_tool.cc

```diff
@@ -502,6 +563,503 @@ void ChunkScrub::chunk_scrub_common()
+  cout << "new iteration thread: " << n <<std::endl;
+
+  vector<AioCompletion*> completions;
+  for (ObjectCursor current_object = shard_start;
+    current_object < shard_end;) {
+    std::vector<ObjectItem> objects;
+    // Get the list of object IDs to deduplicate
+    std::tie(objects, current_object) = get_objects(
+      current_object,
+      shard_end,
+      100);
+
+    // Pick few objects to be processed. Crawling mode decides how many objects
+    // to pick (sampling ratio). Lower sampling ratio makes crawler to has lower
+    // crawling overhead but finds less duplication.
+    std::set<size_t> sampled_indexes = sample_object(objects.size());
+    for (size_t index : sampled_indexes) {
+      ObjectItem target = objects[index];
+      // Only process dirty objects which are expected not processed yet
+      if (is_dirty(target)) {
+        try_dedup_and_accumulate_result(target, completions);
+      }
+    }
+  }
+  for (auto& duplicable_chunk : duplicable_chunks) {
+    mark_dedup(duplicable_chunk, completions);
+  }
+  for (auto& completion : completions) {
+    completion->wait_for_complete();
+  }
+  completions.clear();
+  for (auto& oid : oid_for_evict) {
+    ObjectReadOperation op_tier;
+    AioCompletion* completion_tier = rados.aio_create_completion();
+    if (debug) {
+      cout << "evict " << oid << std::endl;
+    }
+    op_tier.tier_evict();
+    io_ctx.aio_operate(
+        oid,
+        completion_tier,
+        &op_tier,
+        NULL);
+    completions.push_back(completion_tier);
+  }
+  for (auto& completion : completions) {
+    completion->wait_for_complete();
+  }
+  cout << "done iteration thread: " << n <<std::endl;
+}
+
+bool SampleDedup::is_dirty(ObjectItem& object) {
+  ObjectReadOperation op;
+  bool dirty = false;
+  int r = -1;
+  op.is_dirty(&dirty, &r);
+  io_ctx.operate(object.oid, &op, NULL);
+  return dirty;
+}
+
+void SampleDedup::prepare_rados() {
+  int ret = rados.init_with_context(g_ceph_context);
+  if (ret < 0) {
+     cerr << "couldn't initialize rados: " << cpp_strerror(ret) << std::endl;
+     throw std::exception();
+  }
+  ret = rados.connect();
+  if (ret) {
+     cerr << "couldn't connect to cluster: " << cpp_strerror(ret) << std::endl;
+     throw std::exception();
+  }
+}
+
+std::tuple<ObjectCursor, ObjectCursor> SampleDedup::get_shard_boundary() {
+  ObjectCursor shard_start;
+  ObjectCursor shard_end;
+  io_ctx.object_list_slice(begin, end, n, m, &shard_start, &shard_end);
+
+  return std::make_tuple(shard_start, shard_end);
+}
+
+std::tuple<std::vector<ObjectItem>, ObjectCursor> SampleDedup::get_objects(
+  ObjectCursor current, ObjectCursor end, size_t max_object_count) {
+  std::vector<ObjectItem> objects;
+  ObjectCursor next;
+  int ret = io_ctx.object_list(
+    current,
+    end,
+    max_object_count,
+    {},
+    &objects,
+    &next);
+  if (ret < 0 ) {
+    cerr << "error object_list : " << cpp_strerror(ret) << std::endl;
+    throw std::exception();
```

Comment:
how can the crawler handle this exception? I don't see the error handler.

Reply:
added